File: src/Core/LotMatching/LotMatchingMethods/ProvidedLotMatchingMethod.py
from typing import Sequence

from Core.FinancialEvents.Schemas.ProcessedGenericFormats import (
    GenericTaxLot,
    GenericTradeEvent,
)
from Core.LotMatching.Contracts.LotMatchingMethod import LotMatchingMethod
from Core.LotMatching.Schemas.Lot import Lot, LotAcquired, LotSold
from Core.LotMatching.Schemas.Trade import Trade


class ProvidedLotMatchingMethod(LotMatchingMethod):
    def __init__(self, predefinedLots: Sequence[GenericTaxLot[GenericTradeEvent, GenericTradeEvent]]) -> None:
        super().__init__()
        self.predefinedLots: Sequence[GenericTaxLot[GenericTradeEvent, GenericTradeEvent]] = predefinedLots

    def performMatching(self, events: Sequence[Trade]) -> Sequence[Lot]:

        lots = self.predefinedLots

        for lot in lots:
            acquiredTrade = Trade(ID=lot.Acquired.ID, Quantity=lot.Acquired.ExchangedMoney.UnderlyingQuantity, Date=lot.Acquired.Date)
            self.tradeAssociationTracker.trackAcquiredQuantity(acquiredTrade, lot.Quantity)

            # NOTE: Lots subtract, as sells remove from holding
            soldTrade = Trade(ID=lot.Sold.ID, Quantity=lot.Sold.ExchangedMoney.UnderlyingQuantity, Date=lot.Sold.Date)
            self.tradeAssociationTracker.trackSoldQuantity(soldTrade, -lot.Quantity)

        processedLots: list[Lot] = []
        for lot in lots:
            acquiredTrade = Trade(ID=lot.Acquired.ID, Quantity=lot.Acquired.ExchangedMoney.UnderlyingQuantity, Date=lot.Acquired.Date)
            soldTrade = Trade(ID=lot.Sold.ID, Quantity=lot.Sold.ExchangedMoney.UnderlyingQuantity, Date=lot.Sold.Date)

            acquired = LotAcquired(Date=acquiredTrade.Date, Relation=acquiredTrade)
            sold = LotSold(Date=soldTrade.Date, Relation=soldTrade)
            processedLot = Lot(Quantity=lot.Quantity, Acquired=acquired, Sold=sold)
            processedLots.append(processedLot)

        return processedLots

    def generateTradesFromLotsWithTracking(self, lots: Sequence[Lot]) -> Sequence[Trade]:

        processedTrades: dict[str, Trade] = dict()

        for lot in lots:
            acquiredTrade = lot.Acquired.Relation

            acquiredTradeTracking = self.tradeAssociationTracker.getAcquiredTradeTracker(acquiredTrade)
            accountedForAcquiredQuantity = acquiredTradeTracking.Quantity

            acquiredTradeProcessed = Trade(ID=acquiredTrade.ID, Quantity=accountedForAcquiredQuantity, Date=acquiredTrade.Date)
            processedTrades[acquiredTrade.ID] = acquiredTradeProcessed

            soldTrade = lot.Sold.Relation
            soldTradeTracking = self.tradeAssociationTracker.getSoldTradeTracker(soldTrade)
            accountedForSoldQuantity = soldTradeTracking.Quantity

            soldTradeProcessed = Trade(ID=soldTrade.ID, Quantity=accountedForSoldQuantity, Date=soldTrade.Date)
            processedTrades[soldTrade.ID] = soldTradeProcessed

        processedTradesValues = list(processedTrades.values())

        processedTradesValues.sort(key=lambda trade: trade.Date)

        return processedTradesValues

File: lib/Core/LotMatching/ProvidedLotMatchingMethod.ts
import type { Lot } from "./Lot.ts";
import type { LotMatchingMethod } from "./LotMatchingMethod.ts";
import type { Trade } from "./Trade.ts";
import { TradeAssociationTracker } from "./TradeAssociationTracker.ts";
import type { TaxLot } from "@brrr/Core/Schemas/Lots.ts";
import type { TradeEvent } from "@brrr/Core/Schemas/Events.ts";

export class ProvidedLotMatchingMethod implements LotMatchingMethod {
	private tradeAssociationTracker: TradeAssociationTracker = new TradeAssociationTracker();
	private predefinedLots: TaxLot<TradeEvent, TradeEvent>[];

	constructor(predefinedLots: TaxLot<TradeEvent, TradeEvent>[]) {
		this.predefinedLots = predefinedLots;
	}

	performMatching(_events: Trade[]): Lot[] {
		this.tradeAssociationTracker = new TradeAssociationTracker();

		const lots = this.predefinedLots;

		for (const lot of lots) {
			const acquiredTrade: Trade = {
				id: lot.acquired.id,
				quantity: lot.acquired.exchangedMoney.underlyingQuantity,
				date: lot.acquired.date,
			};
			this.tradeAssociationTracker.trackAcquiredQuantity(acquiredTrade, lot.quantity);

			const soldTrade: Trade = {
				id: lot.sold.id,
				quantity: lot.sold.exchangedMoney.underlyingQuantity,
				date: lot.sold.date,
			};
			this.tradeAssociationTracker.trackSoldQuantity(soldTrade, -lot.quantity);
		}

		const processedLots: Lot[] = [];
		for (const lot of lots) {
			const acquiredTrade: Trade = {
				id: lot.acquired.id,
				quantity: lot.acquired.exchangedMoney.underlyingQuantity,
				date: lot.acquired.date,
			};
			const soldTrade: Trade = {
				id: lot.sold.id,
				quantity: lot.sold.exchangedMoney.underlyingQuantity,
				date: lot.sold.date,
			};

			processedLots.push({
				quantity: lot.quantity,
				acquired: { date: acquiredTrade.date, relation: acquiredTrade },
				sold: { date: soldTrade.date, relation: soldTrade },
			});
		}

		return processedLots;
	}

	generateTradesFromLotsWithTracking(lots: Lot[]): Trade[] {
		const processedTrades = new Map<string, Trade>();

		for (const lot of lots) {
			const acquiredTrade = lot.acquired.relation;
			const acquiredTracking = this.tradeAssociationTracker.getAcquiredTradeTracker(acquiredTrade);
			processedTrades.set(acquiredTrade.id, {
				id: acquiredTrade.id,
				quantity: acquiredTracking.quantity,
				date: acquiredTrade.date,
			});

			const soldTrade = lot.sold.relation;
			const soldTracking = this.tradeAssociationTracker.getSoldTradeTracker(soldTrade);
			processedTrades.set(soldTrade.id, {
				id: soldTrade.id,
				quantity: soldTracking.quantity,
				date: soldTrade.date,
			});
		}

		return [...processedTrades.values()].sort((a, b) => a.date.toMillis() - b.date.toMillis());
	}
}
